scripts/shtf.py

```diff
@@ -0,0 +1,629 @@
+        breadcrumb = self.notified_state.get(str(request_id))
+        if breadcrumb is None:
+            self.notified_state[str(request_id)] = dict()
+            self.notified_state[str(request_id)][notified_on] = True
+        else:
+            breadcrumb[notified_on] = True
+
+    def persist(self) -> None:
+        with open(os.path.join(os.path.dirname(__file__), "breadcrumbs.json"), "w") as f:
+            json.dump(self.notified_state, f)
+
+
+@dataclass
+class Context:
+    stats: TransferStats = field(default_factory=TransferStats)
+    messages: list = field(default_factory=list)
+    agent_address: str = ""
+
+    notification_state: NotificationState = field(default_factory=NotificationState)
+
+    def add_message(self, message: NotificationType) -> None:
+        self.messages.append(message)
+
+
+class Notify:
+    def __init__(self, system: str, config: dict[str, dict[str, str]]) -> None:
+        self.system = system
+        self.config = config
+
+    def send(self, message: MessageType, callback: Optional[Callable] = None) -> None:
+        global notifications_sent
+        if notifications_sent > 0 and notifications_sent % 5 == 0:
+            time.sleep(timeout_between_sends)
+
+        text = message["text"]
+
+        if message.get("message_link"):
+            text += "\n" + f"View on [explorer]({message['message_link']})"
+
+        if self.system == "rocketchat":
+            status = self.send_to_rocketchat(text)
+        elif self.system == "telegram":
+            status = self.send_to_telegram(text)
+        else:
+            raise ValueError("Unknown notification system")
+
+        if status:
+            notifications_sent += 1
+            if callback is not None:
+                callback()
+
+    def send_to_rocketchat(self, text: str) -> bool:
+        headers = {"Content-Type": "application/json"}
+
+        body = {
+            "username": "beamer-monitoring-bot",
+            "channel": self.config["rocketchat"]["channel"],
+            "icon": ":heart",
+            "text": text,
+        }
+
+        r = requests.post(self.config["rocketchat"]["url"], data=json.dumps(body), headers=headers)
+
+        if r.status_code == 200:
+            return True
+
+        return False
+
+    def send_to_telegram(self, text: str) -> bool:
+        global notifications_sent
+
+        token = self.config["telegram"]["token"]
+        chat_id = self.config["telegram"]["chat_id"]
+
+        url = f"https://api.telegram.org/bot{token}/sendMessage?chat_id={chat_id}&text={text}"
+        r = requests.get(url).json()
+
+        if r["ok"]:
+            return True
+
+        return False
+
+
+notifications_sent = 0
+timeout_between_sends = 61
+claimRequestExtension = 86400
+global_config: Optional[ConfigType] = None
+
+
+def _default_config() -> dict:
+    return {"chains": {}, "tokens": {}}
+
+
+def set_config(path: Path) -> None:
+    global global_config
```

Comment:
maybe instead of global config you can pass this to other functions/classes in main ?

Reply:
No! 
I need the config deep in some functions and the parent functions don't need the config, so I don't want to pass around config as a parameter.

Reply:
you mean deep copy in a function ? are you changing config in some of them ?

Reply:
No, i don't mean deep copy. 

that analyze_transfers call anyalze_transfer on every transfer and analyze_transfer then calls
```
    check_if_request_created(transfer, ctx)
    check_if_request_has_fill(transfer, ctx)
    check_if_claim_is_made_for_fill(transfer, ctx)
    check_if_challenge_game(transfer, ctx)
    check_if_request_has_been_withdrawn(transfer, ctx)
```

and eventually on eof this will call create_expired_request_message which will call link_to_explorer which will then need the config. So I find it useless to pass the config between 5 functions only to be used by the one deepest in the stack.